services/app-api/src/resolvers/healthPlanPackage/unlockHealthPlanPackage.ts

```diff
@@ -63,89 +76,242 @@ export function unlockHealthPlanPackageResolver(
+                if (contractResult instanceof NotFoundError) {
+                    const errMessage = `A package must exist to be unlocked: ${pkgID}`
+                    logError('unlockHealthPlanPackage', errMessage)
+                    setErrorAttributesOnActiveSpan(errMessage, span)
+                    throw new UserInputError(errMessage, {
+                        argumentName: 'pkgID',
+                    })
+                }
 
-        const pkg: HealthPlanPackageType = result
-        const pkgStatus = packageStatus(pkg)
-        const currentRevision = pkg.revisions[0]
+                const errMessage = `Issue finding a package. Message: ${contractResult.message}`
+                logError('unlockHealthPlanPackage', errMessage)
+                setErrorAttributesOnActiveSpan(errMessage, span)
+                throw new GraphQLError(errMessage, {
+                    extensions: {
+                        code: 'INTERNAL_SERVER_ERROR',
+                        cause: 'DB_ERROR',
+                    },
+                })
+            }
 
-        // Check that the package is in an unlockable state
-        if (pkgStatus === 'UNLOCKED' || pkgStatus === 'DRAFT') {
-            const errMessage = 'Attempted to unlock package with wrong status'
-            logError('unlockHealthPlanPackage', errMessage)
-            setErrorAttributesOnActiveSpan(errMessage, span)
-            throw new GraphQLError(errMessage, {
-                extensions: {
-                    code: 'INTERNAL_SERVER_ERROR',
+            const contract: ContractType = contractResult
+
+            if (contract.draftRevision) {
+                const errMessage = `Attempted to unlock package with wrong status`
+                logError('unlockHealthPlanPackage', errMessage)
+                setErrorAttributesOnActiveSpan(errMessage, span)
+                throw new UserInputError(errMessage, {
+                    argumentName: 'pkgID',
                     cause: 'INVALID_PACKAGE_STATUS',
-                },
-            })
-        }
+                })
+            }
 
-        // pull the current revision out to unlock it.
-        const formDataResult = toDomain(currentRevision.formDataProto)
-        if (formDataResult instanceof Error) {
-            const errMessage = `Failed to decode proto ${formDataResult}.`
-            logError('unlockHealthPlanPackage', errMessage)
-            throw new GraphQLError(errMessage, {
-                extensions: {
-                    code: 'INTERNAL_SERVER_ERROR',
-                    cause: 'PROTO_DECODE_ERROR',
-                },
+            // unlock all the revisions, then unlock the contract, in a transaction.
+            const currentRateRevIDs = contract.revisions[0].rateRevisions.map(
+                (rr) => rr.id
+            )
+            const unlockRatePromises = []
+            for (const rateRevisionID of currentRateRevIDs) {
+                const resPromise = store.unlockRate({
```

Comment:
Do we want to abort all rate unlocks if one fails, then return an error to client? If so, I don't think this will do it. Each one of the unlocks would have to be in the same transaction.

Reply:
you are quite right. I am going to make a ticket for the cleanup to figure this out. Right now we are only doing transactions inside of postgres functions. We need to decide on a design here. Either we should copy your pattern for update where we have a combined "unlockContractAndRates" function that we call from the handler, or we need to make it so that a transaction can be created in a resolver and passed into multiple postgres functions. I'm not sure how to do the latter yet. 

WE'll also need to clean up Submit in the same manner.